storagemarket/impl/providerstates/provider_states.go

```diff
@@ -170,9 +172,20 @@ func EnsureProviderFunds(ctx fsm.Context, environment ProviderDealEnvironment, d
 		return ctx.Trigger(storagemarket.ProviderEventNodeErrored, xerrors.Errorf("looking up miner worker: %w", err))
 	}
 
-	mcid, err := node.EnsureFunds(ctx.Context(), deal.Proposal.Provider, waddr, deal.Proposal.ProviderCollateral, tok)
+	requiredFunds, err := environment.DealFunds().Reserve(deal.Proposal.ProviderCollateral)
+	if err != nil {
+		return ctx.Trigger(storagemarket.ProviderEventTrackFundsFailed, xerrors.Errorf("reserving deal funds: %w", err))
+	}
+
+	mcid, err := node.EnsureFunds(ctx.Context(), deal.Proposal.Provider, waddr, requiredFunds, tok)
 
 	if err != nil {
+		_, err2 := environment.DealFunds().Release(deal.Proposal.ProviderCollateral)
```

Comment:
again this needs to happen in the global fail state handler (`FailDeal`) -- same comment as client